New Leo Layout System

tbpassin opened the pull request:
Implements a new layout system:
- [x] Adds a new _LayoutCacheWidget_ class to cache unused splitters and widgets for reuse;
- [x] Dynamic Window  hosts an instance of _LayoutCacheWidget_;
- [x]  _LayoutCacheWidget_ contains methods to create layouts from layout descriptors;
- [x] New file _plugins/qt_layout.py_ contains _LayoutCacheWidget_ and all the current built-in layouts;
- [x] All layouts can be applied without restarting the outline;
- [x] The default layout specified in the applicable @settings tree can be re-applied without restarting the outline.
- [x] The _DynamicWindow_ startup sequence that creates the initial layout is unchanged;
- [x] `g.app.gui.find_widgets_by_name()` is fixed so it no longer fails to find some widgets;
- [x] Some VR3 commands are adjusted to work better with this new system.
- [ ] For layouts that can take either VR or VR3, generalize the layout structure to allow user to choose which of these alternatives to use in a layout.
- [ ] Decide if VR/VR3 should be visible when a layout that contains them is applied.

qt_layout.py contains minibuffer commands to create all the layouts.  They all start with "layout-".  For the time being, layouts that expect to have VR or VR3 in one of the frames create the frame but do not actually open VR/VR3. They can be displayed in their frame by "vrx-toggle" or "vrx-show" commands.

A companion outline containing a custom menu with many of the new commands, and other layout commands, will be supplied separately.

This PR satisfies Issue https://github.com/leo-editor/leo-editor/issues/4076.

LewisNeal commented:
I tried the new Layout-Demo.leo (Implementation 2 (production)) but some of the commands are not working for me:

_no command function for 'layout-big-tree'
no command function for 'layout-restore-default'
no command function for 'layout-fallback-layout'_

The Make-Body-Tabbed command works.

I ran the Layout-Demo.leo as standalone file and also copied it into myLeoSettings with same result.

tbpassin commented:
It will only work right when you are running the branch I linked to:
```
Leo 6.8.2-devel, tbp-leo_layouts branch, build f03b4e1be1
2024-09-30 23:43:07 -0400
```

 If you were already running that branch, I'm somewhat at a loss.  I (thought that I) had moved over all the commands from my own myLeoSettings file but maybe I missed something.  It's also possible that one or another menu item depended on an obsolete command I had removed.  I tested it on a Linux VM that (I thought) didn't have any of those issues. Perhaps I was wrong about that.
 
 I will double check it all, and if I find anything missing I will put it in and update the version in my repo.

Thanks for testing!

tbpassin commented:
I think I know what happened. I pushed the branch twice.  If you used the first push, you could have gotten those errors. Please update to the current version of the tbp-leo_layouts branch.  I think that will take care of everything.

There is one error message you might still see: `no command function for 'vr3-toggle'` That happened if VR3 has not yet been instantiated for the first time.  That's because it is a method of the VR3 widget. Once VR3 has been made visible by some other command that error message will not appear, and VR3 will toggle.  I knew about this, and I'm mulling over how I want to handle the situation.

LewisNeal commented:
How do I instruct git to use the tbp-leo_layouts branch? I tried git checkout:
$ git checkout f03b4e1
error: pathspec 'f03b4e1' did not match any file(s) known to git
and: 
$ git checkout tbp-leo_layouts
error: pathspec 'tbp-leo_layouts' did not match any file(s) known to git

tbpassin commented:
Get it from my own repo:

https://github.com/tbpassin/leo-editor.git

or download a zip file from

https://github.com/tbpassin/leo-editor/tree/tbp-leo_layouts

LewisNeal commented:
I'm using Git Bash which worked fine with leo-editor (devel)
but not with https://github.com/tbpassin/leo-editor

MINGW64 /n/git (BARE:master)
$ git clone git://github.com/tbpassin/leo-editor/tree/tbp-leo_layouts
Cloning into 'tbp-leo_layouts'...
fatal: repository 'https://github.com/tbpassin/leo-editor/tree/tbp-leo_layouts/' not found

tbpassin commented:
That's the URL for your browser to open my repo in githutb. The address for cloning the whole repo  is (but this will get you everything in the official leo-editor repo and more)

https://github.com/tbpassin/leo-editor.git

Or you can download a zip file of this branch of my repo:

https://github.com/tbpassin/leo-editor/archive/refs/heads/tbp-leo_layouts.zip

This is just an ordinary directory tree, not part of a git repo structure.

LewisNeal commented:
I downloaded the zip file of the branch.
Only comment is that if I select [Restore Default Layout From Settings] there is confirmation in the Log Pane:
 Restoring layout-vertical-thirds layout
but selecting [The Fallback layout] or [Leo's "Legacy" Layout] does not give confirmation in the Log Pane.

Is there a difference between [The Fallback Layout] and [Vertical-thirds Layout] ? 
myLeoSettings file has @string qt-layout-name= vertical-thirds and I can't see any difference.

tbpassin commented:
> Is there a difference between [The Fallback Layout] and [Vertical-thirds Layout] ?

Not that I remember but maybe there's a stray diagnostic statement that I didn't remove.  I'll take a look.

_[Restore Default Layout From Settings]_ looks for the default layout setting, adds the string "layout-" to it, and runs the command by that name.  This approach avoids a name conflict with the layout methods that are already built in.

_[Vertical-thirds Layout]_ constructs the same layout as the current built-in one of that name, only it uses the new techniques.

_[The Fallback Layout]_ is used if there is no default setting to be found. It's the one I like best.

tbpassin commented:
Yes, that message was a diagnostic message that didn't get removed. It was in LeoPyRef, not the Layout-Demo. I've removed it now but it's not worth downloading the updated zip of the branch since that is the only change.

For the Layout-Demo outline, I have slightly updated it.  I removed some obsolete command references that were in the prototype but don't exist anymore.  I don't think they hurt anything.  I cleaned up a few commands, and I removed the menu item "Toggle VR3 In Main Splitter" since the current version of the VR3 command _vr3-toggle_ does the same thing, and probably better. The menu still has a menu item _Toggle VR3_ that dispatches that command.

LewisNeal commented:
Just want to confirm that I like confirmation in the Log Pane:
_Restoring layout-vertical-thirds layout_
It is good feedback to the user. Have you considered a log message when any layout is selected?

tbpassin commented:
> Have you considered a log message when any layout is selected?

Yes, I have.  I wasn't sure if it would be too annoying. OTOH, one wouldn't normally change layouts very often, so maybe not too annoying.  It would be easy enough to make happen.

One weakness of this layout system is that I don't see a good way to query for the current layout.  It would be easy to set a variable to hold it, but it would also be easy to someone to write code to change to a custom layout that didn't update the variable.  In fact, the current Layout-Demo outline contains a mix of layouts - some created by the new system and a few whose command is code that uses parts of the new system to do some layout changes that aren't easy to name and and save.

On a different note, Jacob has expressed an interest in being able to swap objects. e.g., VR3,  between splitters.  This wouldn't be hard, unless one wanted a UI that would be general.  That user interface part might be hard.   Do you have any thoughts about this?

edreamleo commented:
@tbpassin What is the status of this PR?

tbpassin commented:
> @tbpassin What is the status of this PR?

I was hoping you would try it out (and anyone else who has shown interest, like Jacob). Otherwise I think it's merge-worthy.  LewisNeal seems to be all right with it.  I would continue to add layouts and features after a merge. I have worked with it on Linux using a default myLeoSettings.leo so I don't fool myself by calling commands I define in my regular myLeoSettings.  So far not a hitch.

One think we need to agree on is how to present the commands for the six built-in layouts (and probably a few more) to the users.  I favor a menu.  They could be added to the existing Window menu, or by adding a new Layout menu.

I'm open to suggestions.  If we merge without settling this question, everything will continue to work as it has been, including restarting the outline to change the layout.  But a user who knows they are there will be able to use the new commands instead.

I'd rather give them a menu from the start.  If you agree, I'll add one to LeoSettings.leo before the merge.

Another reason I'd like this code to get merged is that I want to make some unrelated improvements to VR3 and I don't want to do that and then find them undone by merging this layout code with its own changes to VR3.

edreamleo commented:
>I was hoping you would try it out (and anyone else who has shown interest, like Jacob). Otherwise I think it's merge-worthy. 

@tbpassin Great!  I'll give the PR a quick look and then merge it into "devel". You can then handle the last two unchecked items in another PR if necessary.

tbpassin commented:
If you want to try it out, get the latest version of the Layout-Demo outline from my clone of leo-editor-contrib. Otherwise you can restore any of the six built-it layouts by prefixing "layout-" to their names.  To see all the new layout commands provided by qt-layout.py , use tab completion on "layout-". The Layout-Demo outline contains some more commands that work with the system but aren't currently provided by qt-layout.py.

edreamleo reviewed (changes requested):
Overall, this PR looks good.  I'll fix a few stylistic changes soon.

edreamleo commented:
@tbpassin I found a few issues in testing. I'll likely push fixes today.

tbpassin commented:
Please don't fix anything without checking with me first.  For one thing I've got some fixes of my own I haven't pushed yet but will later today.  They might take care of what you have found.  For another, I want to make sure that any changes fit smoothly into my overall concept of how things should work (or if you think the concept should be changed, let's discuss that first).

edreamleo commented:
@tbpassin Drat. I fixed lots of mypy errors, and revised the outline structure. I'll have to redo that work when you are done.

edreamleo commented:
@tbpassin I think the best plan is to merge this PR as is. I can then reorganize and tweak `qt_layout.py` as needed.

edreamleo reviewed (approved):
I am good with this PR as a starting point.

I'll retest in "devel" after the merge.

edreamleo commented:
@tbpassin Please feel free to tweak this PR.  Merge when ready!

tbpassin commented:
I'm OK with merging as is. I do have plans for more changes and I want to wait for Viktor's suggestions before moving ahead with them, but as you said, we can keep going afterwards.

I'm fine with anything needed to make mypy happy, and I will **probably** be good with organization changes if you really want them - but we may end up disagreeing about one or two of them!  I suppose you would like to change some of the marked sections into `@others` whereas I like them when they read like pseudo-code because they help remind me of what the algorithm is supposed to be doing.

Do you want me to push the "Merge" button or will you do it?  I didn't know I had permissions to merge anything...

edreamleo commented:
@tbpassin I'll merge now.

edreamleo commented:
@tbpassin Yikes. There seems to be a syntax/tokenizing error in qt_layout.py. I'll fix it in a new PR.

tbpassin commented:
> There seems to be a syntax/tokenizing error in qt_layout.py

That's strange since the branch has been working successfully. When I ALT-Ged to that line it came up in the middle of a docstring. I do see the same error with the current version of my branch, but the stack trace doesn't seem to make sense.

tbpassin commented:
This has got to be a Python parsing bug.  I just ran py_compile on viewrendered3.py and got the same error, and the claimed error was in the line in a triple-quoted docstring:

        VARIABLES USED  See reloadSettings() for the settings' names.

A single apostrophe should be perfectly fine in a triple-quoted string, shouldn't it?  There must be some other quoting problem earlier in the file.

tbpassin commented:
I found the problem.  It was a missing quote char in a triple-quoted string much earlier in the file.  It must have somehow gotten deleted by mistake.  Now the file compiles with py_compile.

tbpassin commented:
What in the heck is going on here? The current version of devel doesn't have my code in it. The cache widget isn't instantiated by DynamicWindow, and qt_layout.py isn't even in LeoPyRef. Since qt_layout isn't imported by DW, its commands are not present.

I thought my branch got merged into devel.  If not, then where is all this work?

```
Leo 6.8.2-devel, devel branch, build ff2dba8bb7
2024-09-27 08:29:29 -0500
```

edreamleo commented:
@tbpassin Please calm down. "devel" *does* have your work:

- LeoPyRef contains a node `@file ../plugins/qt_layout.py`.
- The node `dw.ctor` contains `self.layout_cache = LayoutCacheWidget(c, self)`.
- `Alt-x layout-` shows ten commands.

Did you not pull the latest work to "devel"?

tbpassin commented:
I don't know what happened, but I pulled from devel again and now I got everything. Whew!

Possibly I made a mistake and pulled from my own clone ("origin") instead of from the Leo repo ("upstream"). Origin, of course, hadn't been updated yet since I push updates to it from my computer.  I was so shocked that I didn't think of that for a long time.

Git-Extensions has been a great help to me in working with git and GitHub.  But the difference between pulling from origin and from upstream is one line in a dropdown box.  Maybe that was it, no way to know by now.

edreamleo commented:
@tbpassin I'm glad you have the latest code.
